[REVIEW] ORC writer: add support for column statistics

OlivierNV opened the pull request:
stripe-level and file-level statistics, enabled by default.
resolves #3018 
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+__global__ void __launch_bounds__(128)
+gpuInitStatisticsGroups(statistics_group *groups, const stats_column_desc *cols,
+                        uint32_t num_columns, uint32_t num_rowgroups, uint32_t row_index_stride) {
+  __shared__ __align__(4) volatile statistics_group grp_g[4];
```
What does `grp_g` mean? What is the _g for? Please avoid unnecessary abbreviation.

OlivierNV replied:
A hint that is is shared (global var vs local definition). Convention inherited from other codebases.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+                        uint32_t num_columns, uint32_t num_rowgroups, uint32_t row_index_stride) {
+  __shared__ __align__(4) volatile statistics_group grp_g[4];
+  uint32_t col_id = blockIdx.y;
+  uint32_t ck_id = (blockIdx.x * 4) + (threadIdx.x >> 5);
```
What does `ck` stand for? Please clarify.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+  __shared__ __align__(4) volatile statistics_group grp_g[4];
+  uint32_t col_id = blockIdx.y;
+  uint32_t ck_id = (blockIdx.x * 4) + (threadIdx.x >> 5);
+  uint32_t t = threadIdx.x & 0x1f;
```
I think naming this `lane_id` would be more CUDA idiomatic and clearer.

OlivierNV replied:
I don't. It's also consistent with the rest of cudf-io kernels.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+  uint32_t t = threadIdx.x & 0x1f;
+  volatile statistics_group *grp = &grp_g[threadIdx.x >> 5];
+  if (ck_id < num_rowgroups) {
+    if (!t) {
```
Clearer that only first lane of each warp  updates shared memory:

```suggestion
    if (lane_id == 0) {
```

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+__global__ void __launch_bounds__(128)
+gpuInitStatisticsGroups(statistics_group *groups, const stats_column_desc *cols,
+                        uint32_t num_columns, uint32_t num_rowgroups, uint32_t row_index_stride) {
+  __shared__ __align__(4) volatile statistics_group grp_g[4];
```
Here you hard-code `4` statistics_groups per thread block. But below you decide which threads to use based on `sizeof(uint32_t)` rather than `4`. I suggest instead defining a constexpr variable with a clear name indicating groups per block.

OlivierNV replied:
The number of thread is *not* decided based on sizeof(uint32_t)

harrism replied:
The point is that this code needs either documentation or clarification. A person other than you, or you a few months from now, will not know why this is hard-coded to 4.

OlivierNV replied:
I supposed I could have used 128/32 to make it more obvious where the 4 comes from (both the 128 and 32 constants appears within these 3 consecutive lines of code without having to scroll to see various definitions). Now using a constexpr threads_per_block / threads_per_chunk.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+// blockDim {1024,1,1}
+__global__ void __launch_bounds__(1024, 1)
+gpuInitStatisticsBufferSize(statistics_merge_group *groups, const statistics_chunk *chunks, uint32_t statistics_count) {
+  __shared__ volatile uint32_t scratch_red[32];
```
Use `cudf::experimental::detail::warp_size` instead of `32`.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+    stats_size = 0;
+  }
+  __syncthreads();
+  for (uint32_t start = 0; start < statistics_count; start += 1024) {
```
Use a constexpr variable for block size rather than hard-coding in multiple places.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+      statistics_dtype dtype = col->stats_dtype;
+      switch(dtype) {
+      case dtype_bool8:
+        stats_len = 12 + 1 + 1 + 10;
```
This switch has a lot of hard-coded magic numbers. This makes the code hard to understand and maintain by others (or many weeks later). Please at least document the origin of the numbers, if not define constants and compute the lengths using clear names.

Might also factor this whole switch out into a function.

OlivierNV replied:
Done, but did not move the switch to a separate function since the switch is pretty much the only in its function.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+      }
+    }
+    stats_pos = WarpReducePos32(stats_len, t);
+    if ((t & 0x1f) == 0x1f) {
```
Above comment about avoiding hard-coding warp size and unnecessary optimizations.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+      groups[idx].num_chunks = stats_len;
+    }
+    __syncthreads();
+    if (t == 1023) {
```
Could you use thread 0 instead to make this independent of block size? Also what if `statistics_count < 1023`?

OlivierNV replied:
The last thread of the block has the final sum (it's really mainly a prefix sum). Would have used cub, but it was horribly bad for some reason (4.5K smem and 46+ registers), though perf isn't really critical here...

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+struct stats_state_s {
+  uint8_t *base;              ///< Output buffer start
+  uint8_t *end;               ///< Output buffer end
+  statistics_chunk ck;
```
```suggestion
  statistics_chunk chunk;
```

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+  uint8_t *base;              ///< Output buffer start
+  uint8_t *end;               ///< Output buffer end
+  statistics_chunk ck;
+  statistics_merge_group grp;
```
```suggestion
  statistics_merge_group group;
```

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+  statistics_merge_group grp;
+  stats_column_desc col;
+  // ORC stats
+  uint64_t numberOfValues;
```
```suggestion
  uint64_t num_values;
```

OlivierNV replied:
No. The 'numberOfValues' spelling choice here is to reflect its definition in the ORC protobuf documentation (important for future maintenance by someone who is going to be familiar with the ORC specification)

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+  stats_column_desc col;
+  // ORC stats
+  uint64_t numberOfValues;
+  uint8_t hasNull;
```
```suggestion
  uint8_t has_null;
```

OlivierNV replied:
Same here (protobuf field). Though I think I ended up not writing the hasNull boolean field, as it wasn't being written by other ORC writers (can probably be removed ~ will check)

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+};
+
+
+// Protobuf varint encoding for unsigned int
```
I think we could improve the maintainability of this code by linking to the definition/docs of any protocols we use. In this case, is this the appropriate doc page for protobuf varint encoding? 
https://developers.google.com/protocol-buffers/docs/encoding#varints

Please link appropriate docs for each protobuf function.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+ **/
+// blockDim {128,1,1}
+__global__ void __launch_bounds__(128)
+gpuEncodeStatistics(uint8_t *blob_bfr, statistics_merge_group *groups, const statistics_chunk *chunks,
```
```suggestion
gpu_encode_statistics(uint8_t *blob_buffer, statistics_merge_group *groups, const statistics_chunk *chunks,
```

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+gpuEncodeStatistics(uint8_t *blob_bfr, statistics_merge_group *groups, const statistics_chunk *chunks,
+                    uint32_t statistics_count) {
+  __shared__ __align__(8) stats_state_s state_g[4];
+  uint32_t t = threadIdx.x & 0x1f;
```
As above (warp size dependencies)

OlivierNV replied:
I'll try to make it look like the warp size is abstracted. I don't think it's entirely a coincidence that the warp size matches the register size (to me, 'abstracting' the warp size just makes it more difficult to understand what's really happening)

harrism replied:
FWIW, G80 had a mode where the warp size was only 16 threads (before it was called a warp). It was never exposed, but it is a precedent.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+    case dtype_int16:
+    case dtype_int32:
+    case dtype_int64:
+      // IntegerStatistics = 2
```
Like the one above, this switch has a lot of magic numbers. Please document / clarify.

What does `IntegerStatistics = 2` (and the similar comments in the other cases below) signify?

OlivierNV replied:
The protobuf field id. Perhaps I'll just paste the ORC protobuf definitions in a big comment section so the reader doesn't have to look at the ORC documentation...

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+ **/
+// blockDim {1024,1,1}
+__global__ void __launch_bounds__(1024, 1)
+gpuInitStatisticsBufferSize(statistics_merge_group *groups, const statistics_chunk *chunks, uint32_t statistics_count) {
```
```suggestion
gpu_init_statistics_buffer_size(statistics_merge_group *groups, const statistics_chunk *chunks, uint32_t statistics_count) {
```

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+ **/
+// blockDim {128,1,1}
+__global__ void __launch_bounds__(128)
+gpuInitStatisticsGroups(statistics_group *groups, const stats_column_desc *cols,
```
```suggestion
gpuInit_statistics_groups(statistics_group *groups, const stats_column_desc *cols,
```

harrism commented on cpp/src/io/orc/orc_gpu.h:
```diff
@@ -357,6 +358,49 @@ cudaError_t InitDictionaryIndices(DictionaryChunk *chunks, uint32_t num_columns,
+ *
+ * @return cudaSuccess if successful, a CUDA error code otherwise
+ **/
+cudaError_t OrcEncodeStatistics(uint8_t *blob_bfr, statistics_merge_group *groups, const statistics_chunk *chunks,
```
```suggestion
cudaError_t orc_encode_statistics(uint8_t *blob_bfr, statistics_merge_group *groups, const statistics_chunk *chunks,
```

OlivierNV commented:
omg ::eyeroll::

harrism commented:
@rgsl888prabhu this is awaiting your review. 
@OlivierNV any plans to address comments?

OlivierNV commented:
> @OlivierNV any plans to address comments?

Yup, got delayed with high-priority ORC/parquet bug. Will start addressing the cosmetic review items on Monday.

rgsl888prabhu commented:
> @rgsl888prabhu this is awaiting your review.
> @OlivierNV any plans to address comments?

I was just waiting for response to already provided review.

OlivierNV commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,380 @@
+                        uint32_t num_columns, uint32_t num_rowgroups, uint32_t row_index_stride) {
+  __shared__ __align__(4) volatile statistics_group grp_g[4];
+  uint32_t col_id = blockIdx.y;
+  uint32_t ck_id = (blockIdx.x * 4) + (threadIdx.x >> 5);
```
To clarify: it does not depend on warp size, but on the arbitrary choice of using 32 threads per chunk (which happens to match the warp size).

OlivierNV commented:
@mharris I think I addressed all the review comments. Did not add the dependency on cudf::experimental::detail::warp_size, as even though 32 was chosen as the number of threads per chunk, the kernels themselves do not directly depend on the warp size.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -197,6 +201,23 @@ __device__ inline uint8_t *pb_put_fixed64(uint8_t *p, uint32_t id, const void *r
  * @param[in,out] chunks Statistics data
  * @param[in] statistics_count Number of statistics buffers
  *
+ * ORC statistics format from https://orc.apache.org/specification/ORCv1/
```
Thank you, this is a very helpful addition to the code.

harrism commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -0,0 +1,461 @@
+ * @param[in] row_index_stride Rowgroup size in rows
+ *
+ **/
+constexpr unsigned int log2_init_thread_per_group = 5;
```
Why bother with the log2? Just `init_thread_per_group = 32` and `init_groups_per_block = 4` would be fine. Then you can use multiply and divide instead of shifts. Fewer constexpr variables, shorter names, clearer code.

OlivierNV replied:
It makes it clear that it must be a power of 2

harrism replied:
It's a matter of maintainability and reducing technical debt.

OlivierNV replied:
[Affirming the consequent](https://en.wikipedia.org/wiki/Affirming_the_consequent)

OlivierNV commented on cpp/src/io/orc/stats_enc.cu:
```diff
@@ -42,17 +42,16 @@ gpu_init_statistics_groups(statistics_group *groups, const stats_column_desc *co
       group->num_rows = min(num_rows - min(chunk_id * row_index_stride, num_rows), row_index_stride);
-      __threadfence_block();
-    }
-    SYNCWARP();
```
I just went block-level so as to replace __syncwarp() with __syncthreads() ~ should have no measurable impact on perf here (very low utilization to begin with). Lot smaller change than adding a whole bunch of dependencies on cudf::experimental::detail to get a compile-time warpSize definition :)

OlivierNV commented:
@harrism Update the PR to use a 2D block shape so we don't have to pay the present development cost of arguing about hypthetical maintenance costs of logical vs arithmetic operators.
